pkgs/build-support/rust/fetch-cargo-vendor-util.py

```diff
@@ -0,0 +1,247 @@
+import functools
+import glob
+import hashlib
+import json
+import multiprocessing as mp
+import os
+import re
+import shutil
+import subprocess
+import sys
+
+import requests
+import tomli
+
+
+eprint = functools.partial(print, file=sys.stderr)
+
+
+def load_toml(path: str) -> dict:
+    with open(path, "rb") as f:
+        return tomli.load(f)
+
+
+def download_file_with_checksum(url: str, destination_path: str, expected_checksum: str):
+    """Downloads a file and computes its SHA-256 checksum while writing it to disk."""
+    sha256_hash = hashlib.sha256()
+    with requests.get(url, stream=True) as response:
+        if not response.ok:
+            raise Exception(f"Failed to download file from {url}. Status code: {response.status_code}")
+        with open(destination_path, "wb") as file:
+            for chunk in response.iter_content(1024):  # Download in chunks
+                if chunk:  # Filter out keep-alive chunks
+                    file.write(chunk)
+                    sha256_hash.update(chunk)
+
+    # Compute the final checksum
+    calculated_checksum = sha256_hash.hexdigest()
+
+    if calculated_checksum != expected_checksum:
+        raise Exception(f"Hash mismatch! File fetched from {url} had checksum {calculated_checksum}, expected {expected_checksum}.")
+
+
+def get_download_url_for_tarball(pkg: dict):
+    # TODO: support other registries
+    #       maybe fetch config.json from the registry root and get the dl key
+    #       See: https://doc.rust-lang.org/cargo/reference/registry-index.html#index-configuration
+    if pkg["source"] != "registry+https://github.com/rust-lang/crates.io-index":
+        raise Exception("Only the default crates.io registry is supported.")
+
+    return f"https://crates.io/api/v1/crates/{pkg["name"]}/{pkg["version"]}/download"
+
+
+def download_tarball(pkg: dict, out_dir: str):
+
+    url = get_download_url_for_tarball(pkg)
+    filename = f"{pkg["name"]}-{pkg["version"]}.tar.gz"
+
+    # TODO: allow legacy checksum specification, see importCargoLock for example
+    #       also, don't forget about the other usage of the checksum
+    expected_checksum = pkg["checksum"]
+
+    tarball_out_dir = os.path.join(out_dir, "tarballs", filename)
+    eprint(f"Fetching {url} -> tarballs/{filename}")
+
+    download_file_with_checksum(url, tarball_out_dir, expected_checksum)
+
+
+def download_git_tree(url: str, git_sha_rev: str, out_dir: str):
+
+    tree_out_dir = os.path.join(out_dir, "git", git_sha_rev)
+    eprint(f"Fetching {url}#{git_sha_rev} -> git/{git_sha_rev}")
+
+    subprocess.check_output(["nix-prefetch-git", "--builder", "--quiet", "--url", url, "--rev", git_sha_rev, "--out", tree_out_dir])
```

Comment:
No need to capture output here:

```suggestion
    subprocess.run(["nix-prefetch-git", "--builder", "--quiet", "--url", url, "--rev", git_sha_rev, "--out", tree_out_dir], check=True)
```